File: insert_beginning_dll.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

def insert_at_beginning(head, data):
    newn = Node(data)
    newn.next = head
    if head:
        head.prev = newn
    return newn

def print_list(head):
    cur = head
    while cur:
        print(cur.data, end=" ")
        cur = cur.next
    print()

if __name__ == "__main__":
    head = None
    head = insert_at_beginning(head, 30)
    head = insert_at_beginning(head, 20)
    head = insert_at_beginning(head, 10)
    print("Doubly linked list after inserts at beginning:", end=" ")
    print_list(head)

File: insert_beginning_dll.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *prev, *next;
} Node;

Node* insertAtBeginning(Node* head, int data) {
    Node* newn = (Node*)malloc(sizeof(Node));
    newn->data = data;
    newn->prev = NULL;
    newn->next = head;
    if (head) head->prev = newn;
    return newn;
}

void printList(Node* head) {
    Node* cur = head;
    while (cur) {
        printf("%d ", cur->data);
        cur = cur->next;
    }
    printf("\n");
}

int main() {
    Node* head = NULL;
    head = insertAtBeginning(head, 30);
    head = insertAtBeginning(head, 20);
    head = insertAtBeginning(head, 10);
    printf("Doubly linked list after inserts at beginning: ");
    printList(head);
    return 0;
}
